### DISCUSSION TITLE
Advent of Code - Day 8: Handheld Halt

### DISCUSSION THREAD

**Discussion by e3-cerruti** [e3-cerruti/AoC2020#2] [General]
**Advent of Code - Day 8: Handheld Halt**

Advent of Code continues into day 8 with a fun challenge that has you simulate a simple computer with a single accumulator. For students that can handle the programming this would be a great way to introduce the concepts of assembly and machine language.

Heads up! In past AoC future problems tend to build on the simulator adding complexity. You may consider how you would refactor your code to extract the simulator into a separate class or package.

Not an AP CSA topic, but this example is also a good one to use to demonstrate that if statements are old fashioned. If we foresee our instruction set growing we might recognize that complex if statements and even switch statements could get clumsy.

This is a nice clean switch solution:

```
    switch (command) {
      case "nop" :
        return cur+1;
      case "acc" :
        acc += Integer.parseInt(inp.substring(4));
        return cur+1;
      case "jmp" :
        return cur+Integer.parseInt(inp.substring(4));
    }
```

This solution, also scraped from a forum member but slightly modified shows equivalent if/else:

```
    if(nextOperation.equals("nop"))
    {
       position++;
    }
    else if(nextOperation.equals("acc"))
    {
       accumulator += list.get(position).getArgument();
       position++;
    }
    else
    {
       position += list.get(position).getArgument();
    }
```

With only three instructions these solutions look and work great. But if our machine gets any more complex we start to get a lot of complexity buried in this one method. Wouldn’t it be nice if we were to put the operations in their own methods and then called them?

```
    private void acc(String[] operands) {
       accumulator += Integer.parseInt(operands[0]);
       programCounter += 1;
    }
```

And rather than a complex control structure, what if we use a data structure instead. Let’s try it using a map:

```
    private final Map<String, Consumer<String[]>> instructionSet = new HashMap<>();

    public HandheldGameConsole() {
       instructionSet.put("nop", this::nop);
       instructionSet.put("acc", this::acc);
       instructionSet.put("jmp", this::jmp);
    }
```

Now, when we want to run an instruction we no longer need an if/else block or a switch statement, we can simply execute the correct function from the map.

```
    instructionSet.get(operation).accept(operands);
```

This also demonstrates the power of object oriented programming, if I want to create a new machine with a few additional instructions, I can simply extend this machine and add those instructions to the map. I can also override existing functionality.

There is probably a better way to initialize the instruction set than shown here.

**Comment by zamansky** [comment]
That's the same insight I pointed out in my video/post - function lookup tables. I didn't know how to do it in Java. In Python, you just assign the function to a variable. Similar in Clojure
